tutorials/application_design/py/subscriber.py

```diff
@@ -45,71 +38,110 @@ def __init__(
 
-    async def run(self):
-        await asyncio.gather(
-            self._display_app(),
-            self._metrics_app(),
-            self._transit_app(),
+        new_position_condition = dds.ReadCondition(
+            self._transit_reader, dds.DataState.new_data, new_position_handler
         )
 
-    @property
-    def online_vehicles(self):
-        return {
-            handle: data
-            for handle, data in self._dashboard_data.items()
-            if not data.is_historical
-        }
-
-    @property
-    def offline_vehicles(self):
-        return {
-            handle: data
-            for handle, data in self._dashboard_data.items()
-            if data.is_historical
-        }
-
-    async def _display_app(self):
-        while True:
-            print(f"[[ DASHBOARD: {datetime.now()} ]]")
-            print(f"Online vehicles: {len(self.online_vehicles)}")
-            for data in self.online_vehicles.values():
-                print(f"- Vehicle {data.vin}:")
-                print(f"  Fuel updates: {len(data.fuel_history)}")
-                print(f"  Last known destination: {data.current_destination}")
-                print(f"  Last known fuel level: {data.fuel_history[-1]}")
-            print(f"Offline vehicles: {len(self.offline_vehicles.keys())}")
-            for data in self.offline_vehicles.values():
-                mean_full_consumption = statistics.mean(
-                    data.fuel_history
-                ) / len(data.fuel_history)
-
-                print(f"- Vehicle {data.vin}:")
-                print(f"  Mean fuel consumption: {mean_full_consumption}")
-                print(
-                    f"  Known reached destinations: {len(data.reached_destinations)}"
+        dashboard_condition = dds.GuardCondition()
+
+        def dashboard_handler(_):
+            print(self._dashboard_string())
+
+        dashboard_condition.set_handler(dashboard_handler)
+
+        display_handler_sentinel = threading.Event()
+        display_handler_condition = threading.Condition()
+
+        def display_handler():
+            while not display_handler_sentinel.is_set():
+                with display_handler_condition:
+                    display_handler_condition.wait(5)
+                    dashboard_condition.trigger_value = True
+
+        display_thread = threading.Thread(target=display_handler)
+
+        waitset = dds.WaitSet()
+        waitset.attach_condition(new_position_condition)
+        waitset.attach_condition(dashboard_condition)
+
+        display_thread.start()
+        try:
+            while True:
+                waitset.dispatch()
+                dashboard_condition.trigger_value = False
+        except KeyboardInterrupt:
+            with display_handler_condition:
+                display_handler_sentinel.set()
+                display_handler_condition.notify_all()
+
+        display_thread.join()
+
+    def _new_position_string(self, condition):
```

Comment:
Would it make more sense to have the conditions as instance variables (in `self`) instead of as arguments? Like the readers, they can be set up in the `__init__` method and won't change.

Reply:
I left them as local variables because they do not need to live within `self` and is much easier to explain step-by-step in the documentation.